providers/implementations/digests/digestcommon.c

```diff
@@ -38,6 +38,62 @@ int ossl_digest_default_get_params(OSSL_PARAM params[], size_t blksz,
         ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
         return 0;
     }
+    p = OSSL_PARAM_locate(params, OSSL_DIGEST_PARAM_SECURITY_CATEGORY_COLLISION);
+    if (p != NULL) {
+        int sec_category_collision = 0;
+
+        if (paramsz != 0) {
+            if (paramsz >= 64)
+                sec_category_collision = 5;
+            else if (paramsz >= 48)
+                sec_category_collision = 4;
+            else if (paramsz >= 32)
+                sec_category_collision = 2;
+            else
+                sec_category_collision = 0;
+        } else {
+            if (blksz >= 168)
+                sec_category_collision = 2;
+            else if (blksz >= 136)
+                sec_category_collision = 5;
+            else
+                sec_category_collision = 0;
+        }
+
+        if (!OSSL_PARAM_set_int(p, sec_category_collision)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate(params, OSSL_DIGEST_PARAM_SECURITY_CATEGORY_PREIMAGE);
+    if (p != NULL) {
+        int sec_category_preimage = 0;
+
+        if (paramsz != 0) {
+            if (paramsz >= 64)
+                sec_category_preimage = 5;
+            else if (paramsz == 48)
+                sec_category_preimage = 5;
+            else if (paramsz >= 32)
+                sec_category_preimage = 5;
+            else if (paramsz >= 28)
+                sec_category_preimage = 3;
+            else
+                sec_category_preimage = 1;
+        } else {
+            if (blksz >= 168)
+                sec_category_preimage = 2;
+            else if (blksz >= 136)
+                sec_category_preimage = 5;
```

Comment:
I think this kind of thing is asking for problems.
Even though I know why, it doesn't make a lot of sense that the longer one is weaker…

Might it be better to pass the strengths in as two additional parameters, perhaps with some kind of fallback to checks like these?  This allows exceptions to be specified directly rather than implicitly via constructs like this.